### PR TITLE
[Deprecate _QQ_ 3] Mark _QQ_ as deprecated and normalize language ini parsing

### PR DESCRIPTION
**Pull request by brianteeman**
This is a redo of #13451 ### Summary of Changes

As discussed in https://github.com/joomla/joomla-cms/issues/13421

Deprecate `_QQ_` and normalize parsing of ini language files.
- creates a new method JLanguageHelper::parseIniFile and marks LanguagesHelper::parseFile (helper of com_languages) as deprecated.
- makes the necessary adaptations

### Testing Instructions

1. Apply patch

2. Check all language strings show as usual. Including the ones with `"_QQ_"`

3. Confirm language debug still works.
Enable debug language in global config and system debug plugin.manually make some parsing errors in the en-GB.ini file and reload the page

4. Confirm language overrrides still works.
For instance, load a language string that uses escaped quotes (ex: JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED) and make some changes in the override. Save it and load it again to check if ok.

5. Confirm quoting in server side and client side still works (ex: add to isis template index.php)
```php
$app->enqueueMessage('Server side Test 1 (click to proceed to Client side Test 3): ' . JText::_('JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED'));
JText::script('JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED');
$this->addScriptDeclaration("alert('Client side Test 1: ' + Joomla.JText._('JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED'));");
$this->addScriptDeclaration("alert('Client side Test 2: " . JText::_('JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED', true) . "');");
$this->addScriptDeclaration("window.addEventListener('click', function () { Joomla.renderMessages({'error': ['Client side Test 3: ' + Joomla.JText._('JLIB_DATABASE_ERROR_DATABASE_UPGRADE_FAILED')]}) });");
```

6. Any other test you could remember

### Documentation Changes Required

None

### CODE REVIEW

**Comment by infograf768** [comment]
Looks OK here for what it is supposed to do (tested all cases)

NOTE to @rdeutz while I think of it.
If I remember well, you are the one who made the patch to accept `\"` in core without triggering a debug error.
At that time, as we were just updating core to accept it (`"_QQ_"` had to be present twice in the value), we forced a double use of it (`\"text\"` is OK but a single `\"` IS triggering a debug error.

This is not necessary anymore for 4.0 where we will force the use of the escaped double quote everywhere.
Can you already think about getting rid of that limitation for 4.0?

**Comment by brianteeman** [comment]
please mark your test - thanks

File: administrator/components/com_languages/models/override.php [hunk 1]
```diff
@@ -105,10 +103,12 @@ public function getItem($pk = null)
 			$result->override = $strings[$pk];
 		}
 
-		$opposite_filename = constant('JPATH_' . strtoupper($this->getState('filter.client') == 'site' ? 'administrator' : 'site'))
+
+		$oppositeFileName = constant('JPATH_' . strtoupper($this->getState('filter.client') == 'site' ? 'administrator' : 'site')) 
+
```
Comment (Quy) [inline administrator/components/com_languages/models/override.php]: Remove blank line

**Comment by ReLater** [comment]
I have tested this item :white_check_mark: successfully on 724b56f1833846cac3ded43516e394eaa1cc729c<br /><br />
Nightly build 3.8.8-dev Wednesday, 09 May 2018 02:03:14 UTC.
Tested with no changes in language files.
Tested after replacement of `"_QQ_"` with `\"` in all language files.
Tested with and without Debug Language mode.
tested in FE and BE.
Tested with de-DE and en-GB.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20321">issues.joomla.org/tracker/joomla-cms/20321</a>.</sub>

File: administrator/components/com_languages/models/override.php [hunk 2]
```diff
@@ -105,10 +103,11 @@ public function getItem($pk = null)
 			$result->override = $strings[$pk];
 		}
 
-		$opposite_filename = constant('JPATH_' . strtoupper($this->getState('filter.client') == 'site' ? 'administrator' : 'site'))
-			. '/language/overrides/' . $this->getState('filter.language', 'en-GB') . '.override.ini';
-		$opposite_strings = LanguagesHelper::parseFile($opposite_filename);
-		$result->both = isset($opposite_strings[$pk]) && ($opposite_strings[$pk] == $strings[$pk]);
+
+		$oppositeFileName = constant('JPATH_' . strtoupper($this->getState('filter.client') == 'site' ? 'administrator' : 'site')) 
+							. '/language/overrides/' . $this->getState('filter.language', 'en-GB') . '.override.ini';
```
Comment (Quy) [inline administrator/components/com_languages/models/override.php]: Too many indents. Do it like line 93? Align `=` on line 110,

File: libraries/src/Language/LanguageHelper.php [hunk 3]
```diff
@@ -407,17 +407,62 @@ public static function getContentLanguages($publishedStates = array(1), $checkIn
 		return $languages;
 	}
 
+	/**
+	 * Parse strings from a language file.
+	 *
+	 * @param   string   $fileName  The language ini file path.
+	 * @param   boolean  $debug     If set to true debug language ini file.
+	 *
+	 * @return  boolean  True if saved, false otherwise.
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php:416]: array

File: libraries/src/Language/LanguageHelper.php [hunk 4]
```diff
@@ -430,7 +475,8 @@ public static function saveToIniFile($filename, array $strings)
 		// Write override.ini file with the strings.
 		$registry = new Registry($strings);
 
-		return \JFile::write($filename, $registry->toString('INI'));
+		return \JFile::write($fileName, $registry->toString('INI'));
+
```
Comment (Quy) [inline libraries/src/Language/LanguageHelper.php]: Delete blank line

**Comment by infograf768** [comment]
Waiting for the small changes pointed by @Quy to mark this as successful

File: libraries/src/Language/LanguageHelper.php (hunk 3 above)
Comment (Quy) [reply libraries/src/Language/LanguageHelper.php:416]: The return value is an `array` and not `boolean`.

**Comment by ReLater** [comment]
I have tested this item :white_check_mark: successfully on 96022c3e441d805a23625f185087975cca57f95b<br /><br />After earlier successful test above this time only code review<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20321">issues.joomla.org/tracker/joomla-cms/20321</a>.</sub>

**Comment by infograf768** [comment]
I have tested this item :white_check_mark: successfully on 96022c3e441d805a23625f185087975cca57f95b<br /><br />looks fine to me. @mbabker please also test before merge.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20321">issues.joomla.org/tracker/joomla-cms/20321</a>.</sub>

**Comment by ghost** [comment]
Ready to Commit after two successful tests.

**Comment by brianteeman** [comment]
thanks

**Comment by andrepereiradasilva** [comment]
Hi
@brianteeman beware hosts without parse_ini_file active will not be able to parse the language files.
Basicly the issue solved with https://github.com/joomla/joomla-cms/pull/15620 and latter with wilh  https://github.com/joomla/joomla-cms/pull/17332 will happen again

**Comment by infograf768** [comment]
@mbabker 
Don't know why, but I was thinking this would be for 4.0
@andrepereiradasilva is damned right. We have to deal with parse_ini as this is a 3.x release.

**Comment by brianteeman** [comment]
this is just a redo of @andrepereiradasilva original pr which needed a new owner. That pr was also was against staging

**Comment by infograf768** [comment]
His PR was done a long time ago ( https://github.com/joomla/joomla-cms/pull/13451 ). The corrections were done after that pr was prepared.
Just look at the #

> Basicly the issue solved with #15620 and latter with wilh #17332 will happen again

**Comment by mbabker** [comment]
Should just be [copying this block](https://github.com/joomla/joomla-cms/blob/6770b31323bf38ee991502f5a0934b279836fe29/libraries/src/Language/Language.php#L832-L846) and replacing [this line](https://github.com/joomla/joomla-cms/blob/2cca84d8b6a4d972748cea5e69033418f9447d86/libraries/src/Language/LanguageHelper.php#L444).

**Comment by infograf768** [comment]
@andrepereiradasilva 

I see the code in Language.php here. 
https://github.com/joomla/joomla-cms/blob/staging/libraries/src/Language/Language.php#L832-L846
Are you sure it is missing somewhere?

**Comment by infograf768** [comment]
OK, this is for 3.9 as 3.8 is fine.
I do not have a 3.9 fork here atm, therefore someone has to create a PR towards 3.9

**Pull request merged**